docs/design/multipart-upload.md

```diff
@@ -0,0 +1,82 @@
+# Multipart Upload
+
+## Abstract
+
+This design document describes multipart upload and how it can be implemented.
+Currently, there is no direct support for multipart upload on the Satellite side.
+The feature is simulated on the gateway side, which requires keeping 
+a lot of data in memory and doesn't achieve the full usefulness of multipart upload.
+
+## Background
+
+Multipart upload gives quite a few nice properties:
+
+* Concurrent uploads,
+* continuation of partial uploads, and
+* out-of-order uploads.
+
+The flow for a multipart upload, from a S3 perspective, is:
+
+1. starts a new multipart upload,
+2. the upload data is split into multiple parts,
+3. for each part: send the part,
+4. calls "complete multipart upload," or
+5. alternatively, calls "cancel multipart upload."
+
+There are a few important things we must keep in mind:
+
+* the parts can have different sizes,
+* during the start of the upload, the part size can be unknown,
+* the parts can be uploaded from different computers,
+* the parts can be uploaded out of order, and
+* the parts can be reuploaded, in case of a failed upload.
+
+## Design
+
+We change the path handling to support listing parts uploaded in arbitrary order. When we start an upload we will create an object at the following path:
+
+```
+<project-id>/<bucket-id>/objects/<path ...>_<version> => object information (partial)
+```
+
+The object will contain a `<stream-id>` that defines the immutable content. We suffix the path with the version number to distinguish between different versions.
+
+We don't know the part sizes, and segments cannot be arbitrarily large, which means we need to split each part into multiple segments. To uniquely find segments we'll assign them a unique number `<part-number>_<segment-number>`, which we call the _segment position_.
```

Comment:
Same here regarding the underscore. Can we define the segment position as `<part-number>/<segment-number>` instead of `<part-number>_<segment-number>`?

Reply:
The segment position would end up as a single `int64`, the `_` is a way of delimiting numbers.

Reply:
How does this make a difference?

If we use `/` instead of `_` then we can list and iterate parts and segments like we do for object and folders.

For example:
- If we want to list all segment of a part, we could list with prefix `<project-id>/<bucket-id>/streams/<stream-id>/<part-number>`
- If we want to count all parts of an object, we could list non-recursively with prefix `<project-id>/<bucket-id>/streams/<stream-id>`

Reply:
We can still do the same by using the stream-id as the prefix. So, list everything that has `<project-id>/<bucket-id>/streams/<stream-id>/<part-number>` as the prefix.

Reply:
Oh, you are saying that the KeyValueStore does not really know anything about path delimiters and the prefix does not need to consist of complete path elements.

In such case, yeah, no need to change anything.

Reply:
Hmm, but what would be the list prefix to list count the number of parts in an object? If we list everything with `<project-id>/<bucket-id>/streams/<stream-id>` as prefix it will return all segments, not just the parts.

Reply:
You would get that by listing all the segments and finding the unique parts... however, true, it could be beneficial to list all the parts. Listing all parts would be important only for uploads, so I'm not too worried about it either. I'll think about it a bit.